pyuoi/decomposition/CUR.py

```diff
@@ -0,0 +1,158 @@
+            RandomState instance used by np.random.
+
+        tol : float, optional
+            Tolerance for ARPACK. 0 means machine precision. Ignored by
+            randomized SVD solver.
+
+        Attributes
+        ----------
+        columns_ : ndarray
+            The indices of the columns selected by the algorithm.
+        """
+        self.n_resamples = n_resamples
+        self.max_k = max_k
+        self.resample_frac = resample_frac
+        self.algorithm = algorithm
+        self.n_iter = n_iter
+        self.tol = tol
+        self.random_state = random_state
+
+    def fit(self, X, c=None, stratify=None):
+        """Performs column subset selection in the UoI framework on a provided
+        matrix.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            The data matrix.
+
+        stratify : array-like or None, default None
+            Ensures groups of samples are alloted to resamples proportionally.
+            Labels for each group must be an int greater than zero. Must be of
+            size equal to the number of samples, with further restrictions on
+            the number of groups.
+
+        Returns
+        -------
+        union : ndarray
+            A numpy array containing the indices of the selected columns.
+        """
+        n_samples, n_features = X.shape
+        # initialize list of lists to contain the extracted column indices
+        indices = [[] for _ in range(self.max_k)]
+
+        # truncated SVD fitter object
+        tsvd = TruncatedSVD(n_components=self.max_k,
+                            algorithm=self.algorithm,
+                            n_iter=self.n_iter,
+                            tol=self.tol,
+                            random_state=self.random_state)
+
+        # iterate over bootstraps
+        for bootstrap in range(self.n_resamples):
+            # extract resample
+            X_train, _ = train_test_split(X, test_size=1 - self.resample_frac,
+                                          stratify=stratify,
+                                          random_state=self.random_state)
+
+            # perform truncated SVD on resample
+            tsvd.fit(X_train)
+            # extract right singular vectors
+            V = tsvd.components_.T
+
+            # iterate over ranks
+            for k_idx, k in enumerate(range(1, self.max_k + 1)):
+                # perform column selection on the subset of singular vectors
+                if c is None:
+                    column_flags = self.column_select(V[:, :k], k + 20)
+                else:
+                    column_flags = self.column_select(V[:, :k], c)
+                # convert column flags to set
+                column_indices = set(np.argwhere(column_flags).ravel())
+                indices[k_idx].append(column_indices)
+
+        # calculate intersections
+        intersection = [set.intersection(*indices[k_idx])
```

Comment:
Do we always want hard intersection here and not something like stability selection?

Reply:
Stability selection could be a future feature.